flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java

```diff
@@ -0,0 +1,697 @@
+                if (frozenNodeToStartNodeMap.containsKey(inEdge.getSourceId())) {
+                    info.addTransitiveInEdge(inEdge);
+                }
+            }
+        }
+    }
+
+    private Map<Integer, OperatorChainInfo> buildAndGetChainEntryPoints(
+            List<StreamNode> streamNodes, JobVertexBuildContext jobVertexBuildContext) {
+        for (StreamNode streamNode : streamNodes) {
+            int streamNodeId = streamNode.getId();
+            if (isSourceChainable(streamNode, streamGraph)) {
+                generateHashesByStreamNodeId(streamNodeId);
+                createSourceChainInfo(streamNode, pendingChainEntryPoints, jobVertexBuildContext);
+            } else {
+                pendingChainEntryPoints.computeIfAbsent(
+                        streamNodeId, ignored -> new OperatorChainInfo(streamNodeId));
+            }
+        }
+        return getChainEntryPoints();
+    }
+
+    private Map<Integer, OperatorChainInfo> getChainEntryPoints() {
+
+        final Map<Integer, OperatorChainInfo> chainEntryPoints = new HashMap<>();
+        Iterator<Map.Entry<Integer, OperatorChainInfo>> iterator =
+                pendingChainEntryPoints.entrySet().iterator();
+
+        while (iterator.hasNext()) {
+            Map.Entry<Integer, OperatorChainInfo> entry = iterator.next();
+            Integer startNodeId = entry.getKey();
+            OperatorChainInfo chainInfo = entry.getValue();
+            if (!isReadyToCreateJobVertex(chainInfo)) {
+                continue;
+            }
+            chainEntryPoints.put(startNodeId, chainInfo);
+            iterator.remove();
+        }
+        return chainEntryPoints;
+    }
+
+    /**
+     * Responds to calculating the forward group and reset the parallelism of all nodes in the same
+     * forward group to make them the same.
+     *
+     * @param sourceNodes the source nodes.
+     */
+    private void setVertexParallelismsForDynamicGraphIfNecessary(List<StreamNode> sourceNodes) {
+        Map<Integer, List<StreamEdge>> transitiveNonChainableOutEdgesMap = new HashMap<>();
+        Map<StreamNode, List<StreamNode>> topologicallySortedChainedStreamNodesMap =
+                new LinkedHashMap<>();
+        Set<Integer> visitedStartNodes = new HashSet<>();
+
+        List<StreamNode> enterPoints =
+                getEnterPoints(sourceNodes, topologicallySortedChainedStreamNodesMap);
+
+        for (StreamNode streamNode : enterPoints) {
+            traverseFullGraph(
+                    streamNode.getId(),
+                    streamNode.getId(),
+                    transitiveNonChainableOutEdgesMap,
+                    topologicallySortedChainedStreamNodesMap,
+                    visitedStartNodes);
+        }
+        computeForwardGroupAndSetNodeParallelisms(
+                transitiveNonChainableOutEdgesMap, topologicallySortedChainedStreamNodesMap);
+    }
+
+    private List<StreamNode> getEnterPoints(
```

Comment:
IIUC, this is the actual `getChainEntryPoints()`? And the existing `getChainEntryPoints()` is `getChainEntryPointsReadyToCreateChain()`?